File: plyushkin-service/src/main/java/com/plyushkin/budget/expense/ExpenseNoteId.java
package com.plyushkin.budget.expense;

import static lombok.AccessLevel.PROTECTED;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

@Embeddable
@NoArgsConstructor(access = PROTECTED)
@EqualsAndHashCode
public class ExpenseNoteId {

  @Column(name = "id")
  private long value;

  public static ExpenseNoteId create(long value) throws InvalidExpenseNoteIdException {
    if (value <= 0) {
      throw new InvalidExpenseNoteIdException(
          "Value should be positive but it is: " + value
      );
    }
    ExpenseNoteId expenseNoteId = new ExpenseNoteId();
    expenseNoteId.value = value;
    return expenseNoteId;
  }

  @Override
  public String toString() {
    return String.valueOf(value);
  }

  public static class InvalidExpenseNoteIdException extends Exception {

    public InvalidExpenseNoteIdException(String message) {
      super(message);
    }
  }
}

File: plyushkin-backend/src/main/kotlin/com/plyushkin/budget/expense/ExpenseNoteId.kt
package com.plyushkin.budget.expense

import arrow.core.Either

data class ExpenseNoteId internal constructor(val value: Long) {
    companion object {}
}

class InvalidExpenseNoteId(message: String) : IllegalArgumentException(message)

fun ExpenseNoteId.Companion.create(value: Long): Either<InvalidExpenseNoteId, ExpenseNoteId> =
    value.let {
        if (it <= 0) {
            Either.Left(InvalidExpenseNoteId("Value should be positive: $value"))
        } else {
            Either.Right(ExpenseNoteId(value))
        }
    }

